tests/test_optimize_qng.py

```diff
@@ -63,10 +63,65 @@ def circuit(params):
         var = np.array([0.011, 0.012])
         opt = qml.QNGOptimizer(stepsize=0.01)
 
-        _, res = opt.step_and_cost(circuit, var)
+        step1, res = opt.step_and_cost(circuit, var)
+        step2 = opt.step(circuit, var)
 
         expected = circuit(var)
+        expected_step = var - opt._stepsize * 4 * qml.grad(circuit)(var)
         assert np.all(res == expected)
+        assert np.allclose(step1, expected_step)
+        assert np.allclose(step2, expected_step)
+
+    def test_step_and_cost_with_grad_fn(self, tol):
+        """Test that the correct cost and update is returned via the step_and_cost
+        method for the QNG optimizer when providing an explicit grad_fn"""
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circuit_grouped_input(params):
+            qml.RX(params[0], wires=0)
+            qml.RY(params[1], wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        @qml.qnode(dev)
+        def circuit_separated_input(params_0, params_1):
+            qml.RX(params_0, wires=0)
+            qml.RY(params_1, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        var = np.array([0.011, 0.012])
+        opt = qml.QNGOptimizer(stepsize=0.01)
+
+        for i, circuit in enumerate((circuit_grouped_input, circuit_separated_input)):
+            if i == 0:
```

Comment:
Nice test! However, this `for` loop and `if` statement implies to me that this should be four separate tests!

Reply:
Good call!